### PR TITLE
added damage done stat and cast effciencies

### PR DESCRIPTION
**Pull request by CollCrom**
Started working on the Ret Paladin parser. Knocking out the easy stuff first will move onto more in depth things when I have more time

### CODE REVIEW

File: .env.local.example [hunk 1]
```diff
@@ -1,2 +0,0 @@
-# Copy this file and rename it to `.env.local` to configure your own WCL API key. You can get your own API key from the bottom of the Warcraft Logs settings: https://www.warcraftlogs.com/accounts/changeuser
```
Comment (MartijnHols) [inline .env.local.example]: This file is required for other users, please revert this change.

File: src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    ...HEALING_ABILITIES_ON_GCD,
+    SPELLS.JUDGMENT_CAST.id,
+    SPELLS.CRUSADER_STRIKE.id,
+    225141, // http://www.wowhead.com/spell=225141/fel-crazed-rage (Draught of Souls)
+    SPELLS.DIVINE_STEED.id,
+    26573, // Consecration
+    SPELLS.BLINDING_LIGHT_TALENT.id,
+    642, // Divine Shield
+    SPELLS.LAY_ON_HANDS.id,
+    SPELLS.BEACON_OF_FAITH_TALENT.id,
```
Comment (MartijnHols) [inline src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js]: I don't think these are Retribution abilities. ;)

File: src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+import Icon from 'common/Icon';
+import { formatPercentage, formatDuration } from 'common/format';
+
+import CoreAlwaysBeCastingHealing from 'Parser/Core/Modules/AlwaysBeCastingHealing';
+
+import StatisticBox, { STATISTIC_ORDER } from 'Main/StatisticBox';
+
+const debug = true;
+
+const HEALING_ABILITIES_ON_GCD = [
```
Comment (MartijnHols) [inline src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js]: You shouldn't need this for a DPS.

File: src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
+import React from 'react';
+
+import SPELLS from 'common/SPELLS';
+import ITEMS from 'common/ITEMS';
+import Icon from 'common/Icon';
+import { formatPercentage, formatDuration } from 'common/format';
+
+import CoreAlwaysBeCastingHealing from 'Parser/Core/Modules/AlwaysBeCastingHealing';
```
Comment (MartijnHols) [inline src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js]: This can be the regular AlwaysBeCasting module since you won't need the healing behaviors.

File: src/Parser/RetributionPaladin/Modules/Features/CooldownTracker.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    {
+      spell: SPELLS.AVENGING_WRATH,
+      summary: [
+        BUILT_IN_SUMMARY_TYPES.HEALING,
+        BUILT_IN_SUMMARY_TYPES.OVERHEALING,
+        BUILT_IN_SUMMARY_TYPES.MANA,
+      ],
+    },
+    {
+      spell: SPELLS.HOLY_AVENGER_TALENT,
```
Comment (MartijnHols) [inline src/Parser/RetributionPaladin/Modules/Features/CooldownTracker.js]: I don't think this is available for Ret.

File: src/Parser/RetributionPaladin/Modules/Features/CooldownTracker.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import SPELLS from 'common/SPELLS';
+
+import CoreCooldownTracker, { BUILT_IN_SUMMARY_TYPES } from 'Parser/Core/Modules/CooldownTracker';
+
+class CooldownTracker extends CoreCooldownTracker {
+  static cooldownSpells = [
+    ...CooldownTracker.cooldownSpells,
+    {
+      spell: SPELLS.AVENGING_WRATH,
```
Comment (MartijnHols) [inline src/Parser/RetributionPaladin/Modules/Features/CooldownTracker.js]: Is Crusade the same spell id?

**Comment by CollCrom** [comment]
Thanks for the feedback. I edited the things you suggested. Please take a look when you get a chance!

File: src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    if (cast.ability.guid !== SPELLS.FLASH_OF_LIGHT.id) {
+      return;
+    }
+    const castTime = cast.timestamp - begincast.timestamp;
+    if (!this.constructor.inRange(castTime, globalCooldown, 50)) { // cast times seem to fluctuate by 50ms, not sure if it depends on player latency, in that case it could be a lot more flexible
+      console.warn(`Expected Flash of Light cast time (${castTime}) to match GCD (${Math.round(globalCooldown)}) @${formatDuration((cast.timestamp - this.owner.fight.start_time) / 1000)}`, this.combatants.selected.activeBuffs());
+    }
+  }
+
+  countsAsHealingAbility(cast) {
```
Comment (MartijnHols) [inline src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js]: This can be removed as it's not relevant for your spec.

File: src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    super.recordCastTime(
+      castStartTimestamp,
+      globalCooldown,
+      begincast,
+      cast,
+      spellId
+    );
+    this.verifyCast(begincast, cast, globalCooldown);
+  }
+  verifyCast(begincast, cast, globalCooldown) {
```
Comment (MartijnHols) [inline src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js]: This can be removed as it's not relevant for your spec.

File: src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    SPELLS.BLINDING_LIGHT_TALENT.id,
+    642, // Divine Shield
+    SPELLS.LAY_ON_HANDS.id,
+    SPELLS.BLESSING_OF_FREEDOM.id,
+    1022, // Blessing of Protection
+    853, // Hammer of Justice
+    SPELLS.HAND_OF_RECKONING.id,
+  ];
+
+  recordCastTime(
```
Comment (MartijnHols) [inline src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js]: This can be removed as it's not relevant for your spec.

File: src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+  countsAsHealingAbility(cast) {
+    const spellId = cast.ability.guid;
+    if (spellId === SPELLS.HOLY_SHOCK_CAST.id && !cast.targetIsFriendly) {
+      debug && console.log(`%cABC: ${cast.ability.name} (${spellId}) skipped for healing time; target is not friendly`, 'color: orange');
+      return false;
+    }
+    return super.countsAsHealingAbility(cast);
+  }
+
+  static inRange(num1, goal, buffer) {
```
Comment (MartijnHols) [inline src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js]: This can be removed as it's not relevant for your spec.

File: src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    return num1 > (goal - buffer) && num1 < (goal + buffer);
+  }
+
+  suggestions(when) {
+    const nonHealingTimePercentage = this.totalHealingTimeWasted / this.owner.fightDuration;
+    const deadTimePercentage = this.totalTimeWasted / this.owner.fightDuration;
+
+    when(nonHealingTimePercentage).isGreaterThan(0.3)
```
Comment (MartijnHols) [inline src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js]: This can be removed as it's not relevant for your spec.

File: src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+    }
+    return super.countsAsHealingAbility(cast);
+  }
+
+  static inRange(num1, goal, buffer) {
+    return num1 > (goal - buffer) && num1 < (goal + buffer);
+  }
+
+  suggestions(when) {
+    const nonHealingTimePercentage = this.totalHealingTimeWasted / this.owner.fightDuration;
```
Comment (MartijnHols) [inline src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js]: This can be removed as it's not relevant for your spec.

File: src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    when(deadTimePercentage).isGreaterThan(0.2)
+      .addSuggestion((suggest, actual, recommended) => {
+        return suggest('Your dead GCD time can be improved. Try to Always Be Casting (ABC); try to reduce the delay between casting spells and when you\'re not healing try to contribute some damage.')
+          .icon('spell_mage_altertime')
+          .actual(`${formatPercentage(actual)}% dead GCD time`)
+          .recommended(`<${formatPercentage(recommended)}% is recommended`)
+          .regular(recommended + 0.15).major(1);
+      });
+  }
+  statistic() {
```
Comment (MartijnHols) [inline src/Parser/RetributionPaladin/Modules/Features/AlwaysBeCasting.js]: This should include a statistic for displaying the dead GCD time, for example:

https://github.com/MartijnHols/WoWAnalyzer/blob/4fcfe5839acc7d929944e57bee5652e1ede58f79/src/Parser/AfflictionWarlock/Modules/Features/AlwaysBeCasting.js#L63-L74

**Comment by CollCrom** [comment]
I changed the things you suggested and fixed the eslint errors. I appreciate your patience with me on this!

**Comment by MartijnHols** [comment]
Looks good, thanks!
